### ISSUE TITLE
Upstream Gradient of variables in tf.custom_gradient is (unexpectedly and seemingly generally) zero

### ISSUE THREAD

**Issue by LJKS** [tensorflow/tensorflow#55296]
**System information**
TF 2.8, via Google Colab (so none of the other information should be relevant here?)

**Describe the current behavior**
Using tf.custom_gradient the upstream of a variable currently seems to be zero generally. Exemplary code:

```
def call_model_with_weights(model, x, weights):
  vars = [var.assign(val) for var, val in zip(model.trainable_weights, weights)]
  y = model(x)
  len_vars = len(weights)
  len_y = len(y)

  #stitching together the gradient
  @tf.custom_gradient
  def copy_gradient(var, val, y):
    def grad(*upstream):
      var_up = upstream[:len_vars]
      y_up = upstream[len_vars:]
      return var_up, var_up, y_up

    return (var, y), grad

  _, _y = copy_gradient(model.trainable_weights, weights, y) #this does not change when changing 'model.trainable_weights' to 'vars'
  return _y

model=some_model
x=some_input
weights = some_list_of_model_weights_with_appropriate_shapes #in the simplest case just model.trainable_weights
with tf.GradientTape() as tape:
  tape.watch(weights)
  y = call_model_with_weights(model, x, weights)
gradient=tape.gradient(y, weights) #this gradient is zero (not even NaN as in not connected, but just appropriate shapes of zeros)
```

(background: Trying to implement an equivalent to PyTorch [higher](https://github.com/facebookresearch/higher) library, to enable meta-learning research also on TF, where it currently strongly lacks behind)

I have come upon this while trying to implement a wrapper that calls a function as a function of weights. I observe that generally, the upstream gradient of any variable seems to be zero.

I'm reporting this as a bug, because the behavior is (1) unexpected, (2) undocumented (from what I could find at least), and (3) the expected behavior would generally be more useful from what I understand. 

A full colab notebook reproducing the problem can be found here: https://colab.research.google.com/drive/1kSQhgtxaIhwwpzp0VObZQbqGC0dmxeNc?usp=sharing

I hope this is not just some other stupid coding mistake I made somewhere, but I've been trying to eliminate any other possible reasons for this behavior I could come up with.

**Describe the expected behavior**
The upstream gradient of a variable in tf.custom_gradient should be the actual upstream gradient, not a vector of zeros of the same shape

- Do you want to contribute a PR? (yes/no): Since I'm not sure how thiis could be fixed, no; If I get pointed in the correct direction I will happily do that! 
- Briefly describe your candidate solution(if contributing): See above

**Standalone code to reproduce the issue**
Link again: https://colab.research.google.com/drive/1kSQhgtxaIhwwpzp0VObZQbqGC0dmxeNc?usp=sharing

**Other info / logs** 
None

**Comment by sushreebarsa** [comment]
@sachinprasadhs  I was able to reproduce the issues on colab using TF v[2.8.0](https://colab.research.google.com/gist/sushreebarsa/b2b115903e074e955da48c03125a59a1/copyfor-github-of-call_model_as_function_of_weights.ipynb#scrollTo=lM94qa2YPzKO) and [tf-nightly](https://colab.research.google.com/gist/sushreebarsa/9a0a49e29822bedc9fd467d7d0fde224/copyfor-github-of-call_model_as_function_of_weights.ipynb#scrollTo=lM94qa2YPzKO) ,please find the attached gists. Thanks!

**Comment by wangpengmit** [comment]
Hi @LJKS, thanks for reporting! I think there are two issues here: (1) TF doesn't support differentiating through variable writes; (2) your workaround doesn't work for its own reason. About (2), because `copy_gradient`'s first return value isn't used, `var_up` in `grad` will also be zeros.

I can't think of a working workaround though. So the current situation is that differentiating through variable writes is impossible in TF. We are exploring some ways to properly support it.

**Comment by Carbandoin** [comment]
TensorFlow is an open [source](https://cncrouterhub.com/best-cnc-router/) software library developed by Google that is widely used for machine learning and deep learning applications. One of the key features of TensorFlow is its ability to create custom gradients for differentiable functions. In this article, we will explore the concept of upstream gradient of variables in tf.custom_gradient and how it can impact the performance of machine learning models.
When working with machine learning models, it is important to understand the concept of gradients. Gradients are used to optimize the weights and biases of a model by minimizing the loss function. TensorFlow provides a mechanism for creating custom gradients using the tf.custom_gradient function. This function allows users to define their own gradients for a given function.

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
